packages/graphiql/src/state/GraphiQLSchemaProvider.tsx

```diff
@@ -0,0 +1,189 @@
+import * as React from 'react';
+import { GraphQLSchema } from 'graphql';
+import {
+  generateActionTypeMap,
+  DispatchWithEffects,
+  useReducers,
+  Reducer,
+} from './useReducers';
+import { defaultSchemaLoader } from './common';
+import { SchemaConfig } from './types';
+
+/**
+ * Initial State
+ */
+
+export type SchemaState = {
+  schema: GraphQLSchema | null;
+  isLoading: boolean;
+  hasError: boolean;
+  error?: string;
+  config: SchemaConfig;
+};
+
+const isDev = window.location.hostname === 'localhost';
+
+const uri = isDev
+  ? 'http://localhost:8080/graphql'
+  : 'https://swapi-graphql.netlify.com/.netlify/functions/index';
+
+export const initialReducerState: SchemaState = {
```

Comment:
Is it possible to use discriminated unions so invalid states are impossible to represent?

A simplified example (Reason syntax since it's terser, but I think TS can represent this):
```
type schemaState = 
 | Loading
 | Error(string)
 | Loaded(schema)
```

Reply:
this is just temporary for the demo, but yes it is, via typescript. that would come in a later PR

Reply:
lets talk about state in the plugin spec